### PR TITLE
Observium

### PR DESCRIPTION
**Pull request by AntonZelenin**

### CODE REVIEW

File: agent/src/agent/pipeline/config/jython_scripts/observium.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+def get_data():
+    return _get(
+        sdc.userParams['URL'],
+        sdc.userParams['PARAMS'],
+        DATA_KEYS[sdc.userParams['ENDPOINT']]
+    )
+
+
+def _transform(dict_, key):
```
Comment (dariakharlan) [inline agent/src/agent/pipeline/config/jython_scripts/observium.py:79]: this function is used one time and has a very generic name, are you sure you need it?

File: agent/src/agent/pipeline/config/jython_scripts/observium.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            return res.json()[response_key]
+        except requests.HTTPError as e:
+            requests.post(sdc.userParams['MONITORING_URL'] + str(res.status_code))
+            sdc.log.error(str(e))
+            if i == N_REQUESTS_TRIES:
+                raise
+            time.sleep(2 ** i)
+
+
+def get_data():
```
Comment (dariakharlan) [inline agent/src/agent/pipeline/config/jython_scripts/observium.py:71]: this function is used one time and has a very generic name, are you sure you need it?

File: agent/src/agent/pipeline/config/jython_scripts/observium.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        obj['sysName'] = devices[obj['device_id']]['sysName']
+        obj['location'] = devices[obj['device_id']]['location']
+    return data
+
+
+def create_metrics(data_):
+    return [transform(datum) for datum in data_.values()]
+
+
+def transform(datum):
```
Comment (dariakharlan) [inline agent/src/agent/pipeline/config/jython_scripts/observium.py:104]: this function is used one time and has a very generic name, are you sure you need it?

File: agent/src/agent/pipeline/config/jython_scripts/observium.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+def get_devices():
+    devices = _get(
+        sdc.userParams['DEVICES_URL'],
+        {},
+        'devices'
+    )
+    return _transform(devices, 'device_id')
+
+
+def add_sys_name_and_location(data):
```
Comment (dariakharlan) [inline agent/src/agent/pipeline/config/jython_scripts/observium.py:92]: can this be done in create_metrics function? Just add it to dimension right away, why do it two times

File: agent/src/agent/pipeline/config/stages/source/observium.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+                {'key': 'DEVICES_URL', 'value': urllib.parse.urljoin(base_url, 'devices')},
+                {'key': 'URL', 'value': urllib.parse.urljoin(base_url, self.pipeline.source.config['endpoint'])},
+                {'key': 'API_USER', 'value': self.pipeline.source.config[source.ObserviumSource.USERNAME]},
+                {'key': 'API_PASSWORD', 'value': self.pipeline.source.config[source.ObserviumSource.PASSWORD]},
+                {'key': 'PARAMS', 'value': self.pipeline.config.get('params', {})},
+                {'key': 'MEASUREMENTS', 'value': list(self.pipeline.values.keys())},
+                {'key': 'DIMENSIONS', 'value': self.pipeline.dimensions},
+                {'key': 'INTERVAL_IN_SECONDS', 'value': str(self.pipeline.interval)},
+                {'key': 'DELAY_IN_SECONDS', 'value': str(self.pipeline.delay)},
+                {'key': 'DAYS_TO_BACKFILL', 'value': self.pipeline.days_to_backfill},
```
Comment (dariakharlan) [inline agent/src/agent/pipeline/config/stages/source/observium.py:29]: this pipeline can't backfill. maybe we shouldn't use it

**Review by dariakharlan — CHANGES_REQUESTED** [review]

File: containers/dummy_destination/main.py [hunk 6]
```diff
@@ -127,5 +127,351 @@ def solarwinds_data_example():
     return json.dumps({'results': []})
 
 
+@app.route('/api/v0/devices', methods=['GET'])
```
[the remark this answers is not in the document]
Comment (AntonZelenin) [reply containers/dummy_destination/main.py:130]: sure, was also thinking about this

File: agent/src/agent/pipeline/json_builder.py [hunk 7]
```diff
@@ -295,6 +295,134 @@ def load(self, client_config):
         return self.client_config
 
 
+class ObserviumLoadClientData(LoadClientData):
```
[the remark this answers is not in the document]
Comment (AntonZelenin) [reply agent/src/agent/pipeline/json_builder.py:298]: for me that would be strange to have all builders in one module, and one in a separate one.
In this case I think we should make a package and move the rest of builders to their own files as well

File: agent/src/agent/pipeline/config/jython_scripts/observium.py (hunk 1 above)
Comment (AntonZelenin) [reply agent/src/agent/pipeline/config/jython_scripts/observium.py:79]: seems I forgot to remove it when the function turned into one line)

File: agent/tests/unit/test_mask_sensitive_data.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+def test_timestamp_validation():
+    dummy_config = {
+        'bla': 'bla-bla',
+        'config': {
+            'bla': 'bla-bla',
+        }
+    }
+    for source_type in source.types.keys():
+        dummy_config['type'] = source_type
+        sensitive_data.mask(dummy_config)
```
Comment (dariakharlan) [inline agent/tests/unit/test_mask_sensitive_data.py:14]: change the funtion name

File: agent/src/agent/pipeline/json_builder/jdbc.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
+from agent.pipeline.json_builder import Builder
+
+
+class JDBCBuilder(Builder):
+    # todo is it correct? used to be ../jdbc
```
Comment (dariakharlan) [inline agent/src/agent/pipeline/json_builder/jdbc.py:5]: remove todo

File: agent/src/agent/pipeline/json_builder/jdbc.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+    VALIDATION_SCHEMA_FILE_NAME = 'jdbc'
+
+
+class JDBCRawBuilder(Builder):
+    # todo is it correct? used to be ../jdbc
```
Comment (dariakharlan) [inline agent/src/agent/pipeline/json_builder/jdbc.py:10]: remove todo

**Review by dariakharlan — CHANGES_REQUESTED** [review]

**Review by dariakharlan — APPROVED** [review]

**Pull request merged**